File: backend/app/services/evaluators/thread_canonical.py
"""Canonical thread-evaluation adapters for persistence, API, and analytics."""

from __future__ import annotations

import re
from typing import Any

from app.services.evaluators.models import normalize_rule_outcome

_RULE_STATUS_PRIORITY = {
    "VIOLATED": 0,
    "FOLLOWED": 1,
    "NOT_APPLICABLE": 2,
    "NOT_EVALUATED": 3,
}

_EFFICIENCY_VERDICTS = {
    "EFFICIENT",
    "ACCEPTABLE",
    "INCOMPLETE",
    "FRICTION",
    "BROKEN",
    "NOT_APPLICABLE",
}

_CORRECTNESS_VERDICTS = {
    "PASS",
    "SOFT_FAIL",
    "HARD_FAIL",
    "CRITICAL",
    "NOT_APPLICABLE",
}

def normalize_efficiency_verdict(raw_verdict: str | None) -> str | None:
    if not raw_verdict:
        return None
    normalized = re.sub(r"[^A-Za-z0-9]+", "_", str(raw_verdict)).strip("_").upper()
    return normalized if normalized in _EFFICIENCY_VERDICTS else None


def normalize_correctness_verdict(raw_verdict: str | None) -> str | None:
    if not raw_verdict:
        return None
    normalized = re.sub(r"[^A-Za-z0-9]+", "_", str(raw_verdict)).strip("_").upper()
    return normalized if normalized in _CORRECTNESS_VERDICTS else None


def _rule_source(
    *,
    source_type: str,
    source_label: str,
    rule_id: str,
    section: str,
    evidence: str,
    status: str,
    followed: bool | None,
) -> dict[str, Any]:
    return {
        "sourceType": source_type,
        "sourceLabel": source_label,
        "ruleId": rule_id,
        "section": section,
        "evidence": evidence,
        "status": status,
        "followed": followed,
    }


def _normalize_rule_outcomes(
    raw_rule_outcomes: list[dict[str, Any]] | None,
    *,
    source_type: str,
    source_label: str,
) -> list[dict[str, Any]]:
    outcomes: list[dict[str, Any]] = []
    for item in raw_rule_outcomes or []:
        if not isinstance(item, dict):
            continue
        rule_id = item.get("rule_id") or item.get("ruleId")
        if not rule_id:
            continue
        status, followed = normalize_rule_outcome(item.get("status"), item.get("followed"))
        outcomes.append(
            _rule_source(
                source_type=source_type,
                source_label=source_label,
                rule_id=rule_id,
                section=item.get("section", ""),
                evidence=item.get("evidence", ""),
                status=status,
                followed=followed,
            )
        )
    return outcomes


def _aggregate_rule_outcomes(
    outcomes: list[dict[str, Any]],
) -> tuple[list[dict[str, Any]], dict[str, int]]:
    grouped: dict[str, list[dict[str, Any]]] = {}
    for outcome in outcomes:
        grouped.setdefault(outcome["ruleId"], []).append(outcome)

    aggregated: list[dict[str, Any]] = []
    summary = {
        "followed": 0,
        "violated": 0,
        "notApplicable": 0,
        "notEvaluated": 0,
        "evaluatedCount": 0,
    }

    for rule_id in sorted(grouped.keys()):
        sources = grouped[rule_id]
        winning = min(
            sources,
            key=lambda item: _RULE_STATUS_PRIORITY.get(item["status"], 99),
        )
        aggregated_item = {
            "ruleId": rule_id,
            "section": next((item.get("section", "") for item in sources if item.get("section")), ""),
            "status": winning["status"],
            "followed": winning["followed"],
            "evidence": winning.get("evidence", ""),
            "sources": sources,
        }
        aggregated.append(aggregated_item)

        if winning["status"] == "FOLLOWED":
            summary["followed"] += 1
            summary["evaluatedCount"] += 1
        elif winning["status"] == "VIOLATED":
            summary["violated"] += 1
            summary["evaluatedCount"] += 1
        elif winning["status"] == "NOT_APPLICABLE":
            summary["notApplicable"] += 1
        else:
            summary["notEvaluated"] += 1

    aggregated.sort(key=lambda item: (_RULE_STATUS_PRIORITY.get(item["status"], 99), item["ruleId"]))
    return aggregated, summary


def _normalize_thread_facts(result: dict[str, Any]) -> dict[str, Any]:
    thread = result.get("thread") or {}
    messages = list(thread.get("messages") or [])
    return {
        "threadId": thread.get("thread_id") or thread.get("threadId") or "",
        "userId": thread.get("user_id") or thread.get("userId") or "",
        "messageCount": thread.get("message_count") or thread.get("messageCount") or len(messages),
        "durationSeconds": thread.get("duration_seconds") or thread.get("durationSeconds") or 0,
        "hasImage": any(bool(message.get("has_image") or message.get("hasImage")) for message in messages),
    }


def _normalize_intent_block(result: dict[str, Any], *, row_intent_accuracy: float | None) -> dict[str, Any]:
    accuracy = result.get("intent_accuracy")
    if accuracy is None:
        accuracy = row_intent_accuracy
    return {
        "accuracy": accuracy,
        "evaluations": list(result.get("intent_evaluations") or []),
    }


def _normalize_efficiency_block(result: dict[str, Any], *, row_efficiency_verdict: str | None) -> dict[str, Any]:
    raw_efficiency = result.get("efficiency_evaluation") or {}
    verdict = normalize_efficiency_verdict(
        raw_efficiency.get("verdict") if isinstance(raw_efficiency, dict) else None
    )
    if verdict is None:
        verdict = normalize_efficiency_verdict(row_efficiency_verdict)
    return {
        "verdict": verdict,
        "taskCompleted": bool(raw_efficiency.get("task_completed")) if isinstance(raw_efficiency, dict) else False,
        "frictionTurns": list(raw_efficiency.get("friction_turns") or []) if isinstance(raw_efficiency, dict) else [],
        "recoveryQuality": raw_efficiency.get("recovery_quality") if isinstance(raw_efficiency, dict) else None,
        "failureReason": (
            raw_efficiency.get("failure_reason")
            or raw_efficiency.get("abandonment_reason")
            or ""
        ) if isinstance(raw_efficiency, dict) else "",
        "reasoning": raw_efficiency.get("reasoning", "") if isinstance(raw_efficiency, dict) else "",
        "ruleOutcomes": _normalize_rule_outcomes(
            raw_efficiency.get("rule_compliance") if isinstance(raw_efficiency, dict) else [],
            source_type="efficiency",
            source_label="Efficiency",
        ),
    }


def _normalize_correctness_block(
    result: dict[str, Any],
    *,
    row_worst_correctness: str | None,
) -> dict[str, Any]:
    normalized_evaluations: list[dict[str, Any]] = []
    raw_evaluations = list(result.get("correctness_evaluations") or [])

    for index, evaluation in enumerate(raw_evaluations):
        if not isinstance(evaluation, dict):
            continue
        normalized_evaluations.append(
            {
                "message": evaluation.get("message") or {},
                "verdict": normalize_correctness_verdict(evaluation.get("verdict")),
                "reasoning": evaluation.get("reasoning", ""),
                "hasImageContext": bool(evaluation.get("has_image_context") or evaluation.get("hasImageContext")),
                "calorieSanity": evaluation.get("calorie_sanity") or evaluation.get("calorieSanity") or {},
                "arithmeticConsistency": evaluation.get("arithmetic_consistency") or evaluation.get("arithmeticConsistency") or {},
                "quantityCoherence": evaluation.get("quantity_coherence") or evaluation.get("quantityCoherence") or {},
                "ruleOutcomes": _normalize_rule_outcomes(
                    evaluation.get("rule_compliance"),
                    source_type="correctness",
                    source_label=f"Correctness #{index + 1}",
                ),
            }
        )

    severity_order = {
        "NOT_APPLICABLE": 0,
        "PASS": 1,
        "SOFT_FAIL": 2,
        "HARD_FAIL": 3,
        "CRITICAL": 4,
    }
    worst_verdict = normalize_correctness_verdict(row_worst_correctness)
    for evaluation in normalized_evaluations:
        verdict = evaluation["verdict"]
        if verdict is None:
            continue
        if worst_verdict is None or severity_order.get(verdict, -1) > severity_order.get(worst_verdict, -1):
            worst_verdict = verdict

    return {
        "worstVerdict": worst_verdict,
        "evaluations": normalized_evaluations,
    }


def build_canonical_thread_evaluation(
    result: dict[str, Any] | None,
    *,
    row_intent_accuracy: float | None = None,
    row_worst_correctness: str | None = None,
    row_efficiency_verdict: str | None = None,
    row_success_status: bool | None = None,
) -> dict[str, Any]:
    payload = dict(result or {})
    existing = payload.get("canonical_thread") or payload.get("canonicalThread")
    if isinstance(existing, dict) and {"facts", "evaluators", "derived"} <= set(existing.keys()):
        return existing
    if {"facts", "evaluators", "derived"} <= set(payload.keys()):
        return payload

    facts = {
        "thread": _normalize_thread_facts(payload),
        "execution": {
            "failedEvaluators": dict(payload.get("failed_evaluators") or payload.get("failedEvaluators") or {}),
            "skippedEvaluators": list(payload.get("skipped_evaluators") or payload.get("skippedEvaluators") or []),
            "hadEvaluationError": bool(payload.get("error") or payload.get("failed_evaluators") or payload.get("failedEvaluators")),
        },
    }

    evaluators = {
        "intent": _normalize_intent_block(payload, row_intent_accuracy=row_intent_accuracy),
        "efficiency": _normalize_efficiency_block(payload, row_efficiency_verdict=row_efficiency_verdict),
        "correctness": _normalize_correctness_block(payload, row_worst_correctness=row_worst_correctness),
        "custom": dict(payload.get("custom_evaluations") or payload.get("customEvaluations") or {}),
    }

    per_source_rule_outcomes = list(evaluators["efficiency"]["ruleOutcomes"])
    for evaluation in evaluators["correctness"]["evaluations"]:
        per_source_rule_outcomes.extend(evaluation["ruleOutcomes"])
    canonical_rule_outcomes, rule_summary = _aggregate_rule_outcomes(per_source_rule_outcomes)

    success_status = payload.get("success_status")
    if success_status is None:
        success_status = row_success_status

    derived = {
        "successStatus": bool(success_status),
        "worstCorrectnessVerdict": evaluators["correctness"]["worstVerdict"],
        "efficiencyVerdict": evaluators["efficiency"]["verdict"],
        "canonicalRuleOutcomes": canonical_rule_outcomes,
        "ruleComplianceSummary": rule_summary,
    }

    return {
        "version": 1,
        "facts": facts,
        "evaluators": evaluators,
        "derived": derived,
    }


def enrich_thread_result_for_api(
    result: dict[str, Any] | None,
    *,
    row_intent_accuracy: float | None = None,
    row_worst_correctness: str | None = None,
    row_efficiency_verdict: str | None = None,
    row_success_status: bool | None = None,
) -> dict[str, Any]:
    enriched = dict(result or {})
    canonical_thread = build_canonical_thread_evaluation(
        enriched,
        row_intent_accuracy=row_intent_accuracy,
        row_worst_correctness=row_worst_correctness,
        row_efficiency_verdict=row_efficiency_verdict,
        row_success_status=row_success_status,
    )
    enriched["canonical_thread"] = canonical_thread
    return enriched

File: src/features/evalRuns/utils/threadCanonical.ts
import type {
  CanonicalCorrectnessThreadEvaluation,
  CanonicalCorrectnessVerdict,
  CanonicalEfficiencyVerdict,
  CanonicalThreadEvaluation,
  CanonicalThreadRuleOutcome,
  CanonicalThreadRuleSource,
  ChatMessage,
  Evaluation,
  EvaluationDetail,
  EvaluationTarget,
  RuleOutcomeStatus,
  ThreadEvalResult,
  ThreadEvalRow,
} from '@/types/evalRuns';
import { getRuleOutcomeStatus } from './ruleCompliance';

const RULE_STATUS_PRIORITY: Record<CanonicalThreadRuleOutcome['status'], number> = {
  VIOLATED: 0,
  FOLLOWED: 1,
  NOT_APPLICABLE: 2,
  NOT_EVALUATED: 3,
};

// Detail.status (PASS|FAIL|NA, the spine vocabulary) → rule outcome status.
const DETAIL_STATUS_TO_RULE: Record<string, RuleOutcomeStatus> = {
  PASS: 'FOLLOWED',
  FAIL: 'VIOLATED',
  NA: 'NOT_APPLICABLE',
};

function toNumber(value: number | string | null | undefined): number | null {
  if (value == null) return null;
  const n = typeof value === 'string' ? Number(value) : value;
  return Number.isFinite(n) ? n : null;
}

function normalizeEfficiencyVerdict(rawVerdict?: string | null): CanonicalEfficiencyVerdict | null {
  const normalized = rawVerdict?.replace(/[^A-Za-z0-9]+/g, '_').replace(/^_+|_+$/g, '').toUpperCase();
  if (
    normalized === 'EFFICIENT'
    || normalized === 'ACCEPTABLE'
    || normalized === 'INCOMPLETE'
    || normalized === 'FRICTION'
    || normalized === 'BROKEN'
    || normalized === 'NOT_APPLICABLE'
  ) {
    return normalized;
  }
  return null;
}

function normalizeCorrectnessVerdict(rawVerdict?: string | null): CanonicalCorrectnessVerdict | null {
  const normalized = rawVerdict?.replace(/[^A-Za-z0-9]+/g, '_').replace(/^_+|_+$/g, '').toUpperCase();
  if (
    normalized === 'PASS'
    || normalized === 'SOFT_FAIL'
    || normalized === 'HARD_FAIL'
    || normalized === 'CRITICAL'
    || normalized === 'NOT_APPLICABLE'
  ) {
    return normalized;
  }
  return null;
}

function ruleSourceFromDetail(
  detail: EvaluationDetail,
  sourceType: CanonicalThreadRuleSource['sourceType'],
  sourceLabel: string,
): CanonicalThreadRuleSource {
  const status = DETAIL_STATUS_TO_RULE[(detail.status ?? '').toUpperCase()] ?? 'NOT_EVALUATED';
  return {
    sourceType,
    sourceLabel,
    ruleId: detail.key,
    section: detail.label ?? undefined,
    evidence: detail.explanation ?? '',
    status,
    followed: status === 'FOLLOWED' ? true : status === 'VIOLATED' ? false : null,
  };
}

function aggregateRuleOutcomes(ruleSources: CanonicalThreadRuleSource[]): CanonicalThreadRuleOutcome[] {
  const grouped = new Map<string, CanonicalThreadRuleSource[]>();
  for (const source of ruleSources) {
    grouped.set(source.ruleId, [...(grouped.get(source.ruleId) ?? []), source]);
  }

  return Array.from(grouped.entries())
    .map(([ruleId, sources]) => {
      const winning = [...sources].sort((left, right) => RULE_STATUS_PRIORITY[left.status] - RULE_STATUS_PRIORITY[right.status])[0];
      return {
        ruleId,
        section: sources.find((source) => source.section)?.section,
        evidence: winning.evidence,
        status: winning.status,
        followed: winning.status === 'FOLLOWED' ? true : winning.status === 'VIOLATED' ? false : null,
        sources,
      };
    })
    .sort((left, right) => {
      const byStatus = RULE_STATUS_PRIORITY[left.status] - RULE_STATUS_PRIORITY[right.status];
      return byStatus !== 0 ? byStatus : left.ruleId.localeCompare(right.ruleId);
    });
}

function findEvaluation(target: EvaluationTarget, name: string): Evaluation | undefined {
  return target.evaluations.find((e) => e.evaluatorRef?.name === name);
}

function ruleDetails(evaluation: Evaluation | undefined): EvaluationDetail[] {
  return (evaluation?.details ?? []).filter((d) => d.style === 'rule');
}

/**
 * Build the frozen thread view-model from the structured spine target.
 *
 * The spine carries the evaluator verdicts + rule-compliance atoms; the rich
 * provenance the spine does not store (per-message transcript, intent
 * evaluations, friction turns, recovery quality, reasoning, custom outputs,
 * sanity sub-checks) is read from the raw ``result`` it travels with.
 */
function buildCanonicalFromSpine(
  target: EvaluationTarget,
  result: ThreadEvalResult | undefined,
  row?: ThreadEvalRow,
): CanonicalThreadEvaluation {
  const messages = result?.thread?.messages ?? [];

  const correctnessEval = findEvaluation(target, 'correctness');
  const efficiencyEval = findEvaluation(target, 'efficiency');
  const intentEval = findEvaluation(target, 'intent');

  const efficiencyRuleOutcomes = ruleDetails(efficiencyEval)
    .map((detail) => ruleSourceFromDetail(detail, 'efficiency', 'Efficiency'));

  // The spine carries correctness rule atoms on one evaluation (no per-message
  // split). The raw result still holds the per-message correctness rows the
  // table renders; rule outcomes are sourced from the spine to keep the
  // aggregate identical to the persisted atoms.
  const rawCorrectness = result?.correctness_evaluations ?? [];
  const correctnessEvaluations: CanonicalCorrectnessThreadEvaluation[] = rawCorrectness.map((evaluation, index) => ({
    message: evaluation.message,
    verdict: normalizeCorrectnessVerdict(evaluation.verdict),
    reasoning: evaluation.reasoning,
    hasImageContext: evaluation.has_image_context,
    calorieSanity: evaluation.calorie_sanity,
    arithmeticConsistency: evaluation.arithmetic_consistency,
    quantityCoherence: evaluation.quantity_coherence,
    ruleOutcomes: (evaluation.rule_compliance ?? [])
      .map((rule): CanonicalThreadRuleSource => {
        const status = getRuleOutcomeStatus({ status: rule.status, followed: rule.followed ?? null });
        return {
          sourceType: 'correctness',
          sourceLabel: `Correctness #${index + 1}`,
          ruleId: rule.rule_id,
          section: rule.section,
          evidence: rule.evidence ?? '',
          status,
          followed: status === 'FOLLOWED' ? true : status === 'VIOLATED' ? false : null,
        };
      })
      .filter((rule) => Boolean(rule.ruleId)),
  }));

  const correctnessRuleSources = ruleDetails(correctnessEval)
    .map((detail) => ruleSourceFromDetail(detail, 'correctness', 'Correctness'));

  const allRuleOutcomes = [...efficiencyRuleOutcomes, ...correctnessRuleSources];
  const canonicalRuleOutcomes = aggregateRuleOutcomes(allRuleOutcomes);
  const followed = canonicalRuleOutcomes.filter((rule) => rule.status === 'FOLLOWED').length;
  const violated = canonicalRuleOutcomes.filter((rule) => rule.status === 'VIOLATED').length;
  const notApplicable = canonicalRuleOutcomes.filter((rule) => rule.status === 'NOT_APPLICABLE').length;
  const notEvaluated = canonicalRuleOutcomes.filter((rule) => rule.status === 'NOT_EVALUATED').length;

  const efficiencyVerdict = normalizeEfficiencyVerdict(
    efficiencyEval?.verdict ?? result?.efficiency_evaluation?.verdict ?? row?.efficiency_verdict ?? null,
  );
  const worstCorrectness = normalizeCorrectnessVerdict(
    correctnessEval?.verdict ?? result?.worst_correctness_verdict ?? row?.worst_correctness ?? null,
  );
  const intentAccuracy = toNumber(intentEval?.headlineScore)
    ?? result?.intent_accuracy
    ?? row?.intent_accuracy
    ?? null;

  const attributes = (target.attributes ?? {}) as Record<string, unknown>;

  return {
    version: 1,
    facts: {
      thread: {
        threadId: target.targetKey || result?.thread?.thread_id || row?.thread_id || '',
        userId: result?.thread?.user_id ?? (typeof attributes.user_id === 'string' ? attributes.user_id : ''),
        messageCount: result?.thread?.message_count ?? messages.length,
        durationSeconds: result?.thread?.duration_seconds ?? 0,
        hasImage: messages.some((message: ChatMessage) => message.has_image),
      },
      execution: {
        failedEvaluators: result?.failed_evaluators ?? {},
        skippedEvaluators: result?.skipped_evaluators ?? [],
        hadEvaluationError: Boolean(result?.error || result?.failed_evaluators)
          || target.evaluations.some((e) => e.status === 'error'),
      },
    },
    evaluators: {
      intent: {
        accuracy: intentAccuracy,
        evaluations: result?.intent_evaluations ?? [],
      },
      efficiency: {
        verdict: efficiencyVerdict,
        taskCompleted: result?.efficiency_evaluation?.task_completed ?? Boolean(result?.success_status ?? row?.success_status),
        frictionTurns: result?.efficiency_evaluation?.friction_turns ?? [],
        recoveryQuality: result?.efficiency_evaluation?.recovery_quality ?? null,
        failureReason: result?.efficiency_evaluation?.failure_reason ?? result?.efficiency_evaluation?.abandonment_reason ?? '',
        reasoning: efficiencyEval?.reasoning ?? result?.efficiency_evaluation?.reasoning ?? '',
        ruleOutcomes: efficiencyRuleOutcomes,
      },
      correctness: {
        worstVerdict: worstCorrectness,
        evaluations: correctnessEvaluations,
      },
      custom: result?.custom_evaluations ?? {},
    },
    derived: {
      successStatus: Boolean(result?.success_status ?? row?.success_status),
      worstCorrectnessVerdict: worstCorrectness,
      efficiencyVerdict,
      canonicalRuleOutcomes,
      ruleComplianceSummary: {
        followed,
        violated,
        notApplicable,
        notEvaluated,
        evaluatedCount: followed + violated,
      },
    },
  };
}

function buildFallbackCanonicalThreadEvaluation(
  result: ThreadEvalResult,
  row?: ThreadEvalRow,
): CanonicalThreadEvaluation {
  const messages = result.thread?.messages ?? [];
  const efficiencyRuleOutcomes = (result.efficiency_evaluation?.rule_compliance ?? [])
    .map((rule) => toCanonicalRuleSource(rule, 'efficiency', 'Efficiency'))
    .filter((rule): rule is CanonicalThreadRuleSource => Boolean(rule));

  const correctnessEvaluations: CanonicalCorrectnessThreadEvaluation[] = (result.correctness_evaluations ?? []).map((evaluation, index) => ({
    message: evaluation.message,
    verdict: normalizeCorrectnessVerdict(evaluation.verdict),
    reasoning: evaluation.reasoning,
    hasImageContext: evaluation.has_image_context,
    calorieSanity: evaluation.calorie_sanity,
    arithmeticConsistency: evaluation.arithmetic_consistency,
    quantityCoherence: evaluation.quantity_coherence,
    ruleOutcomes: (evaluation.rule_compliance ?? [])
      .map((rule) => toCanonicalRuleSource(rule, 'correctness', `Correctness #${index + 1}`))
      .filter((rule): rule is CanonicalThreadRuleSource => Boolean(rule)),
  }));

  const allRuleOutcomes = [
    ...efficiencyRuleOutcomes,
    ...correctnessEvaluations.flatMap((evaluation) => evaluation.ruleOutcomes),
  ];
  const canonicalRuleOutcomes = aggregateRuleOutcomes(allRuleOutcomes);
  const followed = canonicalRuleOutcomes.filter((rule) => rule.status === 'FOLLOWED').length;
  const violated = canonicalRuleOutcomes.filter((rule) => rule.status === 'VIOLATED').length;
  const notApplicable = canonicalRuleOutcomes.filter((rule) => rule.status === 'NOT_APPLICABLE').length;
  const notEvaluated = canonicalRuleOutcomes.filter((rule) => rule.status === 'NOT_EVALUATED').length;

  return {
    version: 1,
    facts: {
      thread: {
        threadId: result.thread?.thread_id ?? row?.thread_id ?? '',
        userId: result.thread?.user_id ?? '',
        messageCount: result.thread?.message_count ?? messages.length,
        durationSeconds: result.thread?.duration_seconds ?? 0,
        hasImage: messages.some((message: ChatMessage) => message.has_image),
      },
      execution: {
        failedEvaluators: result.failed_evaluators ?? {},
        skippedEvaluators: result.skipped_evaluators ?? [],
        hadEvaluationError: Boolean(result.error || result.failed_evaluators),
      },
    },
    evaluators: {
      intent: {
        accuracy: result.intent_accuracy ?? row?.intent_accuracy ?? null,
        evaluations: result.intent_evaluations ?? [],
      },
      efficiency: {
        verdict: normalizeEfficiencyVerdict(result.efficiency_evaluation?.verdict ?? row?.efficiency_verdict ?? null),
        taskCompleted: result.efficiency_evaluation?.task_completed ?? Boolean(result.success_status ?? row?.success_status),
        frictionTurns: result.efficiency_evaluation?.friction_turns ?? [],
        recoveryQuality: result.efficiency_evaluation?.recovery_quality ?? null,
        failureReason: result.efficiency_evaluation?.failure_reason ?? result.efficiency_evaluation?.abandonment_reason ?? '',
        reasoning: result.efficiency_evaluation?.reasoning ?? '',
        ruleOutcomes: efficiencyRuleOutcomes,
      },
      correctness: {
        worstVerdict: normalizeCorrectnessVerdict(result.worst_correctness_verdict ?? row?.worst_correctness ?? null),
        evaluations: correctnessEvaluations,
      },
      custom: result.custom_evaluations ?? {},
    },
    derived: {
      successStatus: Boolean(result.success_status ?? row?.success_status),
      worstCorrectnessVerdict: normalizeCorrectnessVerdict(result.worst_correctness_verdict ?? row?.worst_correctness ?? null),
      efficiencyVerdict: normalizeEfficiencyVerdict(result.efficiency_evaluation?.verdict ?? row?.efficiency_verdict ?? null),
      canonicalRuleOutcomes,
      ruleComplianceSummary: {
        followed,
        violated,
        notApplicable,
        notEvaluated,
        evaluatedCount: followed + violated,
      },
    },
  };
}

function toCanonicalRuleSource(
  rule: { rule_id?: string; section?: string; evidence?: string; status?: CanonicalThreadRuleOutcome['status']; followed?: boolean | null },
  sourceType: CanonicalThreadRuleSource['sourceType'],
  sourceLabel: string,
): CanonicalThreadRuleSource | null {
  if (!rule.rule_id) {
    return null;
  }
  const status = getRuleOutcomeStatus({
    status: rule.status,
    followed: rule.followed ?? null,
  });
  return {
    sourceType,
    sourceLabel,
    ruleId: rule.rule_id,
    section: rule.section,
    evidence: rule.evidence ?? '',
    status,
    followed: status === 'FOLLOWED' ? true : status === 'VIOLATED' ? false : null,
  };
}

export function getCanonicalThreadEvaluation(
  result: ThreadEvalResult,
  row?: ThreadEvalRow,
): CanonicalThreadEvaluation {
  if (result.canonical_thread) return result.canonical_thread;
  if (row?.canonical_thread) return row.canonical_thread;
  const target = row?.target;
  if (target) return buildCanonicalFromSpine(target, result, row);
  return buildFallbackCanonicalThreadEvaluation(result, row);
}
